onnxruntime/core/session/inference_session.cc

```diff
@@ -2299,6 +2299,113 @@ Status InferenceSession::Run(const RunOptions& run_options,
   return retval;
 }
 
+Status InferenceSession::Run(const OrtRunOptions* run_options,
+                             const char* const* input_names,
+                             const OrtValue* const* input, size_t input_len,
+                             const char* const* output_names, size_t output_names_len,
+                             OrtValue** output) {
+  InlinedVector<std::string> feed_names;
+  feed_names.reserve(input_len);
+  InlinedVector<OrtValue> feeds;
+  feeds.reserve(input_len);
+
+  for (size_t i = 0; i != input_len; ++i) {
+    if (input_names[i] == nullptr || input_names[i][0] == '\0') {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "input name cannot be empty");
+    }
+
+    if (!input[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, MakeString("NULL input supplied for input ", input_names[i]).c_str());
+    }
+
+    feed_names.emplace_back(input_names[i]);
+    feeds.emplace_back(*input[i]);
+  }
+
+  // Create output feed
+  InlinedVector<std::string> output_name_vec;
+  output_name_vec.reserve(output_names_len);
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output_names[i] == nullptr || output_names[i][0] == '\0') {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "output name cannot be empty");
+    }
+    output_name_vec.emplace_back(output_names[i]);
+  }
+
+  std::vector<OrtValue> fetches;
+  fetches.reserve(output_names_len);
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output[i] != nullptr) {
+      fetches.emplace_back(*output[i]);
+    } else {
+      fetches.emplace_back();
+    }
+  }
+
+  Status status;
+  if (run_options == nullptr) {
+    OrtRunOptions op;
+    status = Run(op, feed_names, feeds, output_name_vec, &fetches, nullptr);
+  } else {
+    status = Run(*run_options, feed_names, feeds, output_name_vec, &fetches, nullptr);
+  }
+
+  if (!status.IsOK())
+    return status;
+
+  // We do it in two loops to make sure copy __ctors does not throw
```

Comment:
can we std::move them and just use a single loop?

Reply:
That's challenging - we have to rely on a vector of unique ptr as a fail-safe, before filling an array of OrtValue*.

Reply:
I don't quite understand why we need the complexity here. We always allocate new OrtValue for outputs so that they're owned by the caller, and std::move would avoid any potential copy failure. 

And we possibly only need to std::move values what were null in the original `output` given OrtValue contains a shared_ptr so any pre-allocated ones would have been updated directly. 

e.g. something like this: 
```cpp
  for (size_t i = 0; i != output_names_len; ++i) {
    if (output[i] == nullptr) {
      output[i] = new OrtValue(std::move(fetches[i]));
    }
  }
```
Looking at the history that's largely what it used to do (although it copy constructed instead of std::move'd). 

@yuslepukhin can you explain the issue that resulting in adding 2 loops in #14572?

Reply:
IIRC there were 2 reasons: 1. you can't use 'new'. 2. even if you could if an exception is thrown during move/copy of OrtValue it would result in a memory leak.

Reply:
Have we ever seen a failure to copy an output to return to the user resulting in a memory leak? 

If we std::move it there's next to no additional memory required and I would have thought the chance of an exception is virtually non-existent given how basic the OrtValue class is, so mainly wondering if the complexity in the code is justified.

Reply:
Let me try to explain what is happening. 
1) The reason behind two loops is the exception safety, because copy constructor that is employed in the first loop can throw, and the release() in the second loop can not.
2) Said that, if we employ `std::move()` the second loop is not needed because move __ctors are usually and in this case specifically are noexcept.
3) The reason behind copy construction is the semantics of this API. We always allowed the user to retain the copy of the output OrtValue that points to that output buffer, after all they allocated it and they may want to use it to access the buffer. In fact, that is the whole reason behind pre-allocated outputs.

Reply:
Since we already have a copy of user allocated OrtValues, we should do `std::move()` as it is usually noexcept and get rid of the second loop.

Reply:
let's keep the logic as it is, and improve this as separate task.